skills_ml/algorithms/embedding/train.py

```diff
@@ -55,83 +57,110 @@ def __init__(
-                    self._model.build_vocab(batch, update=True)
-
-                self._model.train(batch, total_examples=self._model.corpus_count, epochs=self._model.iter, *args, **kwargs)
-                self.vocab_size_cumu.append(len(self._model.wv.vocab))
-                batch_iter += 1
-                logging.info('\n')
-
-        elif self.model_type == 'doc2vec':
-            corpus_gen = self.corpus_generator
-            reiter_corpus_gen = Reiterable(corpus_gen)
-            self._model.build_vocab(reiter_corpus_gen)
-            self._model.train(reiter_corpus_gen, total_examples=self._model.corpus_count, epochs=self._model.iter, *args, **kwargs)
+        tic = time()
+        try:
+            self.corpus_metadata = corpus_generator.metadata
+        except AttributeError:
+            self.corpus_metadata = None
+
+        if any([model.wv.vocab for model in self._models]):
+            logging.info("Model has been trained")
+            return 0
+
+        for model in self._models:
+            model.model_name = "_".join([
+                model.model_type,
+                self._model_hash(model, self.training_time, self.corpus_metadata)]
+            ) + '.model'
+
+        if self.model_type <= set(['word2vec', 'fasttext']):
+            batch_gen = BatchGenerator(corpus_generator, self.batch_size)
+            if n_processes == 1:
+                for i, batch in enumerate(batch_gen):
+                    logging.info("Training batch #{} ".format(i))
+                    self._models = [self._train_one_batch(model, batch) for model in self._models]
+            else:
+                with mp.Pool(processes=n_processes, maxtasksperchild=1) as pool:
+                    for i, batch in enumerate(batch_gen):
+                        logging.info("Training batch #{} ".format(i))
+                        partial_train = partial(self._train_one_batch, batch=batch, *args, **kwargs)
+                        self._models = pool.map(partial_train, self._models)
+
+        elif set(self.model_type) == set(['doc2vec']):
+            logging.info(f"Training {self.model_type}")
+            reiter_corpus_gen = Reiterable(corpus_generator)
+            self._models = [self._train_one_batch(model, reiter_corpus_gen) for model in self._models]
             if lookup:
-                self.lookup_dict = corpus_gen.lookup
-                self._model.lookup_dict = self.lookup_dict
+                self.lookup_dict = corpus_generator.lookup
+                for model in self._models:
+                    model.lookup_dict = self.lookup_dict
+        else:
+            raise TypeError("Doc2Vec model can only be trained alone with its own kind, not supporting training with other models")
 
-        self._model.model_name = self.model_name
+        logging.info(f"{', '.join([m.model_name for m in self._models])} are trained in {str(timedelta(seconds=time()-tic))}")
+
+    def _model_hash(self, model, training_time, corpus_metadata):
```

Comment:
This could probably just be a standalone function because there are no references to self

Reply:
Or make it so the caller doesn't have to pass in self.training_time or self.corpus_metadata